Sources/NIOSOCKS/Channel Handlers/SOCKSServerHandshakeHandler.swift

```diff
@@ -0,0 +1,120 @@
+                return // do nothing, we've buffered the data
+            }
+            context.fireChannelRead(self.wrapInboundOut(message))
+        } catch {
+            context.fireErrorCaught(error)
+        }
+    }
+    
+    public func handlerAdded(context: ChannelHandlerContext) {
+        do {
+            try self.stateMachine.connectionEstablished()
+        } catch {
+            context.fireErrorCaught(error)
+        }
+    }
+    
+    public func handlerRemoved(context: ChannelHandlerContext) {
+        guard let buffer = self.inboundBuffer else {
+            return
+        }
+        context.fireChannelRead(.init(buffer))
+    }
+    
+    public func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        do {
+            let message = self.unwrapOutboundIn(data)
+            switch message {
+            case .selectedAuthenticationMethod(let method):
+                try self.handleWriteSelectedAuthenticationMethod(method, context: context, promise: promise)
+            case .response(let response):
+                try self.handleWriteResponse(response, context: context, promise: promise)
+            case .authenticationData(let data):
+                try self.handleWriteData(data, context: context, promise: promise)
+            case .authenticationComplete:
+                try self.handleAuthenticationComplete(context: context, promise: promise)
+            }
+        } catch {
+            context.fireErrorCaught(error)
+            promise?.fail(error)
+        }
+    }
+    
+    private func handleWriteSelectedAuthenticationMethod(
+        _ method: SelectedAuthenticationMethod, context: ChannelHandlerContext, promise: EventLoopPromise<Void>?) throws {
+        try stateMachine.sendAuthenticationMethod(method)
+        var buffer = context.channel.allocator.buffer(capacity: 16)
+        buffer.writeMethodSelection(method)
+        context.write(self.wrapOutboundOut(buffer), promise: promise)
+    }
+    
+    private func handleWriteResponse(
+        _ response: SOCKSResponse, context: ChannelHandlerContext, promise: EventLoopPromise<Void>?) throws {
+        try stateMachine.sendServerResponse(response)
+        var buffer = context.channel.allocator.buffer(capacity: 16)
+        buffer.writeServerResponse(response)
+        context.write(self.wrapOutboundOut(buffer), promise: promise)
+    }
+    
+    private func handleWriteData(_ data :ByteBuffer, context: ChannelHandlerContext, promise: EventLoopPromise<Void>?) throws {
+        do {
+            try self.stateMachine.sendData()
+            context.write(self.wrapOutboundOut(data), promise: promise)
+        } catch {
+            promise?.fail(error)
+        }
+    }
+    
+    private func handleAuthenticationComplete(context: ChannelHandlerContext, promise: EventLoopPromise<Void>?) throws {
+        try stateMachine.authenticationComplete()
+        promise?.succeed(())
```

Comment:
We need to issue a zero-length write here. A rule we have about TCP is that sequential write promises must be completed in order, and that requires us to actually issue a write.